# Credit Risk Classification

Credit risk poses a classification problem that’s inherently imbalanced. This is because healthy loans easily outnumber risky loans. In this Challenge, you’ll use various techniques to train and evaluate models with imbalanced classes. You’ll use a dataset of historical lending activity from a peer-to-peer lending services company to build a model that can identify the creditworthiness of borrowers.

## Instructions:

This challenge consists of the following subsections:

* Split the Data into Training and Testing Sets

* Create a Logistic Regression Model with the Original Data

* Predict a Logistic Regression Model with Resampled Training Data 

### Split the Data into Training and Testing Sets

Open the starter code notebook and then use it to complete the following steps.

1. Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

2. Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

    > **Note** A value of `0` in the “loan_status” column means that the loan is healthy. A value of `1` means that the loan has a high risk of defaulting.  

3. Check the balance of the labels variable (`y`) by using the `value_counts` function.

4. Split the data into training and testing datasets by using `train_test_split`.

### Create a Logistic Regression Model with the Original Data

Employ your knowledge of logistic regression to complete the following steps:

1. Fit a logistic regression model by using the training data (`X_train` and `y_train`).

2. Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.

4. Answer the following question: How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Predict a Logistic Regression Model with Resampled Training Data

Did you notice the small number of high-risk loan labels? Perhaps, a model that uses resampled data will perform better. You’ll thus resample the training data and then reevaluate the model. Specifically, you’ll use `RandomOverSampler`.

To do so, complete the following steps:

1. Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

2. Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.
    
4. Answer the following question: How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Write a Credit Risk Analysis Report

For this section, you’ll write a brief report that includes a summary and an analysis of the performance of both machine learning models that you used in this challenge. You should write this report as the `README.md` file included in your GitHub repository.

Structure your report by using the report template that `Starter_Code.zip` includes, and make sure that it contains the following:

1. An overview of the analysis: Explain the purpose of this analysis.


2. The results: Using bulleted lists, describe the balanced accuracy scores and the precision and recall scores of both machine learning models.

3. A summary: Summarize the results from the machine learning models. Compare the two versions of the dataset predictions. Include your recommendation for the model to use, if any, on the original vs. the resampled data. If you don’t recommend either model, justify your reasoning.

In [24]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from imblearn.metrics import classification_report_imbalanced
from sklearn.metrics import confusion_matrix,classification_report
from sklearn.preprocessing import StandardScaler
import warnings
warnings.filterwarnings('ignore')

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [25]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
panda_df = pd.read_csv('./Resources/lending_data.csv')
# YOUR CODE HERE!

# Review the DataFrame
# YOUR CODE HERE!
panda_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [69]:
# Separate the data into labels and features

# Separate the y variable, the labels
# YOUR CODE HERE!]
y_labels = panda_df['loan_status']

In [70]:
# Separate the X variable, the features
# YOUR CODE HERE!
X_features = panda_df.drop(columns=['loan_status'])

In [71]:
# Review the y variable Series
# YOUR CODE HERE!
y_labels.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [72]:
# Review the X variable DataFrame
# YOUR CODE HERE!
X_features.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [56]:
# Check the balance of our target values
# YOUR CODE HERE!


display(y_labels.value_counts())

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [31]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
# YOUR CODE HERE!
X_train, X_test, y_train, y_test = train_test_split(X_features, y_labels, random_state=1)

# 

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [9]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!

lr = LogisticRegression(random_state=1)
lr

# Fit the model using training data
# YOUR CODE HERE!
lr.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [10]:
# Make a prediction using the testing data
# YOUR CODE HERE!
# Make predictions using the test data
y_pred = lr.predict(X_test)
y_pred

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [11]:
results = pd.DataFrame({
    "Prediction": y_pred, 
    "Actual": y_test
}).reset_index(drop=True)


results.head(15)

Prediction  Actual
0            0       0
1            0       0
2            0       0
3            0       0
4            0       0
5            0       0
6            0       0
7            0       0
8            0       0
9            0       0
10           1       1
11           0       0
12           1       1
13           0       0
14           0       0

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [12]:
# Import Modules
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
#from sklearn.datasets.samples_generator import make_blobs
from sklearn.datasets import make_blobs
from sklearn.preprocessing import StandardScaler

In [13]:
# Print the balanced_accuracy score of the model
# YOUR CODE HERE!

print(balanced_accuracy_score(y_test, y_pred))

0.9520479254722232


In [14]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
print(confusion_matrix(y_test, y_pred))

[[18663   102]
 [   56   563]]


In [45]:
# Print the classification report for the model
# YOUR CODE HERE!
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** WRITE YOUR ANSWER HERE!
The logistics regression model is doing what could be interpreted as doing a very decent job at both prediction and recalling both Precision and Recall again for both `0` (healthy loan) & `1` (high-risk loan) with only high-risk loan Precision below 90%.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [33]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

In [34]:
# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!

random_oversampler = RandomOverSampler(random_state=1)


# Fit the original training data to the random_oversampler model
# YOUR CODE HERE!

X_resampled_train, y_resampled_train = random_oversampler.fit_resample(X_train, y_train)


In [55]:

# Count the distinct values of the resampled labels data
# YOUR CODE HERE!display(X_train.value_counts())
display(X_train.value_counts())

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  derogatory_marks  total_debt
9100.0     6.984          46300            0.352052        3                0                 16300         64
9400.0     7.102          47400            0.367089        3                0                 17400         59
9600.0     7.211          48500            0.381443        4                0                 18500         59
9700.0     7.251          48800            0.385246        4                0                 18800         58
9000.0     6.937          45900            0.346405        3                0                 15900         58
                                                                                                            ..
16500.0    10.124         75900            0.604743        10               2                 45900          1
           10.120         75800            0.604222        10               2                 45800          1
16400.0

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [60]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!

logi_model = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
#YOUR CODE HERE!

logi_model.fit(X_resampled_train, y_resampled_train)

# Make a prediction using the testing data
# YOUR CODE HERE!
model_prediction = logi_model.predict(X_test)
model_prediction

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

# Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [61]:
# Print the balanced_accuracy score of the model 
# YOUR CODE HERE!
print(balanced_accuracy_score(y_test, model_prediction))

0.9936781215845847


In [62]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
confusion_matrix(y_test, model_prediction)

array([[18649,   116],
       [    4,   615]], dtype=int64)

In [63]:
# Print the classification report for the model
# YOUR CODE HERE!
print(classification_report(y_test, model_prediction))

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** YOUR ANSWER HERE!
The logistic regression model, fit with the oversampled data predictied both the `0` (healthy loan) and `1` (high-risk loan) labels just sligthy better than our fist model, with the main improvement being the `1` (high-risk loan) recall score incresing by 8 points. 

## Overview of the Analysis

The risk of credit is an inherently imbalanced issue. 
This being due to the number of healthy loans vastly outweighing risky loans. On any given day (other than credit retractive events) you can be use most loan will remain current. With the rise of peer-to-peer lending services (the exact type of data we dealt with for this project) we must be ever more diligent in our data analysis.  This With this task we used a ML technique to train and evaluate models with imbalanced classes and built a model that can identify the creditworthiness of borrowers.

## Results

* Machine Learning Model 1:
•	The resulting balance accuracy score for the first model was 95.2%
•	The Precision score on the Test data for the first model was 1.00 and the Recall was 99%.
•	The Precision score on the Prediction data for the first model was .99 and the Recall was 84%.  
* Machine Learning Model 2:
•	The resulting balance accuracy score for the second model was 0.92%.
•	The Precision score on the Test data for the second model was 1.00 and the Recall was 99%.
•	The Precision score on the Prediction data for the first model was 84 and the Recall was 99%.  

## Summary 

A model is a simplifided version of the observations. The simplifiactions are meant to discard the superflous details that are 


